### DISCUSSION TITLE
Using `templ` for Go code codegen

### DISCUSSION THREAD

**Discussion by danicc097** [a-h/templ#926] [Ideas]
I've been experimenting a bit to replace stdlib go templates with templ for go codegen (not html). The potential is huge but there are some drawbacks I quickly found:
- Escaping is enabled by default, so the output will not be valid go code unless templ.Raw is used, and if I make a custom component implementing templ.Component I will lose LSP help. templ.Raw won't work in expressions like:
  ```go
  for _, lang := range data.Langs {
       { @templ.Raw(fmt.Sprintf(`Lang%s Lang = "%s"`, lang.CamelLang, lang.Lang)) }
  }
  ```
-  Code is harder to read with `templ.Raw( ... )` everywhere and no syntax highlighting like with html.

So in short:
1. Is there a workaround for the escaping issue above?
2. Is it feasible to have an alternate syntax where templ expressions are the ones escaped, in a similar vein to stdlib templates? E.g. the code below (if set to templ language mode in vscode) already has syntax highlighting, and its a matter of finding some expression syntax. A marker comment on top of the `.templ` files could be used for both the extension and the `templ generate` to handle go codegen properly (no escaping for `templ generate`, syntax change in IDE...). This would also open the door to having explicit support to generate other types of files, not just `go|html` (none come to mind now though)

```go
package templates

import "fmt"

templ TranslationCode2(data TemplateData) {
	{{ // templ comments }}
	// Code generated by i18ngo. DO NOT EDIT.
	package translation

	import (
		"bytes"
		"html/template"
	)

	// Translator is implemented by all language translators.
	type Translator interface {
	  {{ // templ comments }}
            {{ for _, msg := range data.Messages }}
              {{ fmt.Sprintf("%s (%s) (string, error)", msg.MethodName, msg.Args) }}
            {{ end }}
	}

	// Lang represents available translated languages.
	type Lang string

	const (
	    {{ // templ comments }}
	  {{ for _, lang := range data.Langs }}
		  {{ fmt.Sprintf(`Lang%s Lang = "%s"`, lang.CamelLang, lang.Lang) }}
	  {{ end }}
	)

	// NewTranslators initializes all translators.
	func NewTranslators() map[Lang]Translator {
		return map[Lang]Translator{
              {{ for _, lang := range data.Langs }}
                {{ fmt.Sprintf(`// writing var %q to comment`, lang.Lang) }}
              {{ end }}
                }
          }
}
```

**Comment by joerdav** [comment]
Hey @danicc097 

We've gone back and forth a few times on this and landed on a focus of HTML at the moment. See https://github.com/a-h/templ/issues/385#issuecomment-1874603038

I think there is demand and a gap in the Go ecosystem for a non-HTML templ, because we have had a few requests and questions around it.

I agree that it would be a great way to do go codegen, but as of now it's not a goal of templ, we want to get HTML right for now :)
